contrib/pg_tde/src/access/pg_tde_xlog_encrypt.c

```diff
@@ -121,121 +128,157 @@ void
+	EncryptionState->segBuf = allocptr;
+
+	pg_atomic_init_u64(&EncryptionState->enc_key_lsn, 0);
+
+	elog(DEBUG1, "pg_tde: initialized encryption buffer %lu bytes", TDEXLogEncryptStateSize());
 }
 
 /*
  * Encrypt XLog page(s) from the buf and write to the segment file.
  */
 static ssize_t
-TDEXLogWriteEncryptedPages(int fd, const void *buf, size_t count, off_t offset)
+TDEXLogWriteEncryptedPages(int fd, const void *buf, size_t count, off_t offset, 
+							TimeLineID tli, XLogSegNo segno)
 {
-	char		iv_prefix[16] = {0,};
-	size_t		data_size = 0;
-	XLogPageHeader curr_page_hdr = &EncryptCurrentPageHrd;
-	XLogPageHeader enc_buf_page = NULL;
-	RelKeyData *key = GetTdeGlobaleRelationKey(GLOBAL_SPACE_RLOCATOR(XLOG_TDE_OID));
-	off_t		enc_off;
-	size_t		page_size = XLOG_BLCKSZ - offset % XLOG_BLCKSZ;
-	uint32		iv_ctr = 0;
+	char iv_prefix[16] = {0,};
+	RelKeyData *key = EncryptionKey;
+	char *enc_buff = EncryptionState->segBuf;
 
 #ifdef TDE_XLOG_DEBUG
-	elog(DEBUG1, "write encrypted WAL, pages amount: %d, size: %lu offset: %ld", count / (Size) XLOG_BLCKSZ, count, offset);
+	elog(DEBUG1, "write encrypted WAL, size: %lu, offset: %ld [%lX], seg: %X/%X, key_start_lsn: %X/%X", 
+					count, offset, offset, LSN_FORMAT_ARGS(segno), LSN_FORMAT_ARGS(key->internal_key.start_lsn));
 #endif
 
-	/*
-	 * Go through the buf page-by-page and encrypt them. We may start or
-	 * finish writing from/in the middle of the page (walsender or
-	 * `full_page_writes = off`). So preserve a page header for the IV init
-	 * data.
-	 *
-	 * TODO: check if walsender restarts form the beggining of the page in
-	 * case of the crash.
-	 */
-	for (enc_off = 0; enc_off < count;)
-	{
-		data_size = Min(page_size, count);
+	SetXLogPageIVPrefix(tli, segno, iv_prefix);
+	PG_TDE_ENCRYPT_DATA(iv_prefix, offset,
+						(char *) buf, count,
+						enc_buff, key);
 
-		if (page_size == XLOG_BLCKSZ)
-		{
-			memcpy((char *) curr_page_hdr, (char *) buf + enc_off, SizeOfXLogShortPHD);
+	return pg_pwrite(fd, enc_buff, count, offset);
+}
 
-			/*
-			 * Need to use a separate buf for the encryption so the page
-			 * remains non-crypted in the XLog buf (XLogInsert has to have
-			 * access to records' lsn).
-			 */
-			enc_buf_page = (XLogPageHeader) (TDEXLogEncryptBuf + enc_off);
-			memcpy((char *) enc_buf_page, (char *) buf + enc_off, (Size) XLogPageHeaderSize(curr_page_hdr));
-			enc_buf_page->xlp_info |= XLP_ENCRYPTED;
-
-			enc_off += XLogPageHeaderSize(curr_page_hdr);
-			data_size -= XLogPageHeaderSize(curr_page_hdr);
-			/* it's a beginning of the page */
-			iv_ctr = 0;
-		}
-		else
-		{
-			/* we're in the middle of the page */
-			iv_ctr = (offset % XLOG_BLCKSZ) - XLogPageHeaderSize(curr_page_hdr);
-		}
 
-		if (data_size + enc_off > count)
-		{
-			data_size = count - enc_off;
-		}
+static RelKeyData *
+pg_tde_create_key_map_entry(const RelFileLocator *newrlocator, uint32 entry_type, XLogRecPtr start_lsn)
```

Comment:
Shouldn't this function go to `tdemap.c`?

Reply:
Yes, makes sene, I'll move it